arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -0,0 +1,464 @@
+	 * cacheable, read-write
+	 * Note: read-write region is marked execute-ever internally
+	 */
+	MMU_REGION_FLAT_ENTRY("zephyr_data",
+			      (uintptr_t)__kernel_ram_start,
+			      (uintptr_t)__kernel_ram_size,
+			      MT_NORMAL | MT_RW | MT_SECURE),
+};
+
+static void setup_page_tables(void)
+{
+	unsigned int index;
+	const struct arm_mmu_region *region;
+	u64_t max_va = 0, max_pa = 0;
+
+	for (index = 0; index < mmu_config.num_regions; index++) {
+		region = &mmu_config.mmu_regions[index];
+		max_va = MAX(max_va, region->base_va + region->size);
+		max_pa = MAX(max_pa, region->base_pa + region->size);
+	}
+
+	__ASSERT(max_va <= (1ULL << CONFIG_ARM64_VA_BITS),
+		 "Maximum VA not supported\n");
+	__ASSERT(max_pa <= (1ULL << CONFIG_ARM64_PA_BITS),
+		 "Maximum PA not supported\n");
+
+	/* create translation tables for user provided platform regions */
+	for (index = 0; index < mmu_config.num_regions; index++) {
+		region = &mmu_config.mmu_regions[index];
+		if (region->size || region->attrs)
+			init_xlat_tables(region);
+	}
+
+	/* setup translation table for zephyr execution regions */
+	for (index = 0; index < ARRAY_SIZE(mmu_zephyr_regions); index++) {
+		region = &mmu_zephyr_regions[index];
+		if (region->size || region->attrs)
+			init_xlat_tables(region);
+	}
+}
+
+static void enable_mmu_el1(unsigned int flags)
+{
+	ARG_UNUSED(flags);
+	unsigned int val;
+
+	/* Set MAIR, TCR and TBBR registers */
+	__asm__ volatile("msr ttbr0_el1, %0"
+			:
+			: "r" ((u64_t)base_xlat_table)
+			: "memory", "cc");
+	__asm__ volatile("msr tcr_el1, %0"
+			:
+			: "r" (get_tcr(1))
+			: "memory", "cc");
+	__asm__ volatile("msr mair_el1, %0"
+			:
+			: "r" (MEMORY_ATTRIBUTES)
+			: "memory", "cc");
+
+	/* Enable the MMU and data cache */
+	__asm__ volatile("mrs %0, sctlr_el1" : "=r" (val));
+	__asm__ volatile("msr sctlr_el1, %0"
+			:
+			: "r" (val | SCTLR_M_BIT | SCTLR_C_BIT)
+			: "memory", "cc");
+
+	/* Ensure the MMU enable takes effect immediately */
+	__ISB();
+
+	MMU_DEBUG("MMU enabled with dcache\n");
+}
+
+/* ARM MMU Driver Initial Setup */
+
+/*
+ * @brief MMU default configuration
+ *
+ * This function provides the default configuration mechanism for the Memory
+ * Management Unit (MMU).
+ */
+static int arm_mmu_init(struct device *arg)
+{
+	unsigned int val, flags = 0;
+	int idx;
+
+	/* Ensure that MMU is already not enabled */
+	__asm__ volatile("mrs %0, sctlr_el1" : "=r" (val));
+	MMU_DEBUG("sctlr_el1: %x\n", val);
+	__ASSERT((val | SCTLR_M_BIT) != 0, "MMU is already enabled\n");
+
+	MMU_DEBUG("xlat tables:\n");
+	MMU_DEBUG("base table(L%d): %p, %d entries\n", XLAT_TABLE_BASE_LEVEL,
+			(u64_t *)base_xlat_table, NUM_BASE_LEVEL_ENTRIES);
+	for (idx = 0; idx < CONFIG_MAX_XLAT_TABLES; idx++)
+		MMU_DEBUG("%d: %p\n", idx, (u64_t *)(xlat_tables + idx));
+
+	setup_page_tables();
+
+	MMU_DEBUG("enabling dcache/mmu\n");
+
+	/* currently only EL1 is supported */
+	enable_mmu_el1(flags);
```

Comment:
We are currently switching to EL1 automatically at boot when `CONFIG_SWITCH_TO_EL1` is set. This is the default case since `CONFIG_SWITCH_TO_EL1` defaults to `=y` but this also means that the MMU is broken if someone sets `CONFIG_SWITCH_TO_EL1=n` and for example we are running in EL3.

Can you add a check before enabling the MMU to check whether we actually are in EL1 and gracefully print an error message otherwise?

Reply:
@carlocaione Thanks for the review.
Current MMU code does not support EL3. I will add print message and gracefully exit if not in EL1.

On a side note, we should be able to support MMU in EL3 by programming EL3 system registers equivalent of EL1 system registers(TTBR_ELx, TCR_ELx, MAIR_ELx, SCTLR_ELx), however, I have not verified it yet.

Reply:
If you can support EL3 even better. Your call.

Reply:
@carlocaione , Like i cited before in your PR, I feel we can add *Zephyr in EL3*  only when there is a real need as this will have maintenance overhead to consider behavior in multiple exception levels and will add code under #ifdefs / if (currentEL == Elx). It will require everyone to run tests both EL3/EL1 support variants for any arch port changes. 
So I feel El3 support can be deferred.

Reply:
>  So I feel El3 support can be deferred.

ACK